server/src/tls.rs

```diff
@@ -0,0 +1,360 @@
+use rustls::{ServerConfig,RootCertStore};
+
+use x509_parser::prelude::{X509Certificate, FromDer};
+use x509_parser::oid_registry::OidRegistry;
+
+use std::fs;
+use std::sync::Arc;
+use std::io::BufReader;
+
+use anyhow::{anyhow, bail, Context, Result};
+use common::encoding::{decode_utf16le, encode_utf16le};
+use common::settings::Collector;
+
+use hyper::body::HttpBody;
+use hyper::{Body, Request};
+use crate::sldc;
+use sha1::{Sha1,Digest};
+use hex::ToHex;
+
+use log::{info,debug};
+
+
+/// Load certificates contained inside a PEM file
+fn load_certs(filename: &str) -> Result<Vec<rustls::Certificate>, std::io::Error> {
+    let certfile = fs::File::open(filename)?;
+    let mut reader = BufReader::new(certfile);
+
+    debug!("Loaded certificate {:?}", filename);
+
+    Ok(rustls_pemfile::certs(&mut reader)?
+        .iter()
+        .map(|v| rustls::Certificate(v.clone()))
+        .collect())
+}
+
+
+/// Load private key contained inside a file
+///
+/// Panics
+///
+/// - When file does not exist or cannot be opened
+/// - When private key is not well-formed PEM private key file
+/// - When the file is well-formed but does not contain any key
+fn load_priv_key(filename: &str) -> rustls::PrivateKey {
+    let keyfile = fs::File::open(filename).expect("Cannot open private key file");
+    let mut reader = BufReader::new(keyfile);
+
+    debug!("Loading private key {:?}", filename);
+
+    loop {
+        match rustls_pemfile::read_one(&mut reader).expect("Cannot parse private key file") {
+            Some(rustls_pemfile::Item::RSAKey(key)) => return rustls::PrivateKey(key),
+            Some(rustls_pemfile::Item::PKCS8Key(key)) => return rustls::PrivateKey(key),
+            Some(rustls_pemfile::Item::ECKey(key)) => return rustls::PrivateKey(key),
+            None => break,
+            _ => {}
+        }
+    }
+
+    panic!(
+        "No keys found in {:?} (encrypted keys not supported)",
+        filename
+    );
+}
+
+
+/// Certificate thumbprint = SHA-1 of entire certificate, as a hexadecimal string
+pub fn compute_thumbprint(cert_content: &[u8]) -> String {
+    let mut hasher = Sha1::new();
+    hasher.update(cert_content);
+    let shasum_it = hasher.finalize();
+
+    shasum_it.encode_hex::<String>()
+}
+
+pub struct ResultConfig {
+    pub server_config : Arc<ServerConfig>,
+    pub thumbprint : String,
+}
+
+
+/// Create configuration for TLS connection
+///
+/// Panics
+///
+/// - When a certificate cannot be added to root of trust
+/// - When the defaults of the configuration could not be setup
+/// - When server certificate/key or CA certificate could not be used for connection configuration
+/// - When CA certificate file does not contain any certificate
+pub fn make_config(args: &common::settings::Tls) -> Result<ResultConfig, std::io::Error> {
+
+    // loading server certificate
+    let cert = match load_certs(args.server_certificate()) {
+        Ok(c) => {c},
```

Comment:
Why the `{}`? It seems you want to use `?`. You should change the "Error" type of `Result` to be able to do this.